File: packages/protocol/typescript/src/session-tabs/values.ts
export const SESSION_TABS_PROTOCOL_CAPABILITY = 'session.tabs.protobuf.v1' as const

export type SessionTabsTabTypeValue = 'terminal' | 'markdown' | 'file' | 'browser'
export type SessionTabsAgentStateValue = 'working' | 'blocked' | 'waiting' | 'done'

// Why: The closed agent union keeps session snapshots compatible with agent launch validation.
export type SessionTabsLaunchAgentValue =
  | 'claude'
  | 'openclaude'
  | 'codex'
  | 'autohand'
  | 'opencode'
  | 'mimo-code'
  | 'pi'
  | 'omp'
  | 'gemini'
  | 'antigravity'
  | 'aider'
  | 'goose'
  | 'amp'
  | 'kilo'
  | 'kiro'
  | 'crush'
  | 'aug'
  | 'cline'
  | 'codebuff'
  | 'command-code'
  | 'continue'
  | 'cursor'
  | 'droid'
  | 'kimi'
  | 'mistral-vibe'
  | 'qwen-code'
  | 'rovo'
  | 'hermes'
  | 'openclaw'
  | 'copilot'
  | 'grok'
  | 'devin'
  | 'ante'
  | 'trae'

export type SessionTabsProviderSessionValue = {
  key: 'session_id' | 'conversation_id'
  id: string
  transcriptPath?: string
}

export type SessionTabsAgentStatusValue = {
  state: SessionTabsAgentStateValue
  prompt: string
  updatedAt: number
  stateStartedAt: number
  paneKey: string
  stateHistory: never[]
  agentType?: string
  model?: string
  tabId?: string
  worktreeId?: string
  connectionId?: string | null
  toolName?: string
  toolInput?: string
  interactivePrompt?: string
  lastAssistantMessage?: string
  interrupted?: boolean
  providerSession?: SessionTabsProviderSessionValue
  promptInteractionKey?: string
}

export type SessionTabsTerminalTabValue =
  | {
      type: 'terminal'
      id: string
      title: string
      parentTabId: string
      leafId: string
      status: 'pending-handle' | 'sleeping'
      terminal: null
      isActive: boolean
      quickCommandLabel?: string | null
      ptyId?: string | null
      color?: string | null
      isPinned?: boolean
      launchAgent?: SessionTabsLaunchAgentValue
      resolvedAgentType?: SessionTabsLaunchAgentValue
      startupCwd?: string
      agentStatus?: SessionTabsAgentStatusValue | null
    }
  | {
      type: 'terminal'
      id: string
      title: string
      parentTabId: string
      leafId: string
      status: 'ready'
      terminal: string
      worktreeInstanceId?: string | null
      isActive: boolean
      quickCommandLabel?: string | null
      ptyId?: string | null
      color?: string | null
      isPinned?: boolean
      launchAgent?: SessionTabsLaunchAgentValue
      resolvedAgentType?: SessionTabsLaunchAgentValue
      startupCwd?: string
      agentStatus?: SessionTabsAgentStatusValue | null
    }

export type SessionTabsMarkdownTabValue = {
  type: 'markdown'
  id: string
  title: string
  filePath: string
  relativePath: string
  language: 'markdown'
  mode: 'edit' | 'markdown-preview'
  isDirty: boolean
  isActive: boolean
  sourceFileId: string
  sourceFilePath: string
  sourceRelativePath: string
  documentVersion: string
  color?: string | null
  isPinned?: boolean
}

export type SessionTabsFileTabValue = {
  type: 'file'
  id: string
  title: string
  filePath: string
  relativePath: string
  language: string
  mode?: 'edit' | 'diff'
  diffSource?: 'staged' | 'unstaged'
  isDirty: boolean
  isActive: boolean
  color?: string | null
  isPinned?: boolean
}

export type SessionTabsBrowserTabValue = {
  type: 'browser'
  id: string
  title: string
  browserWorkspaceId: string
  browserPageId: string | null
  url: string
  loading: boolean
  canGoBack: boolean
  canGoForward: boolean
  isActive: boolean
  color?: string | null
  isPinned?: boolean
}

export type SessionTabsTabValue =
  | SessionTabsTerminalTabValue
  | SessionTabsMarkdownTabValue
  | SessionTabsFileTabValue
  | SessionTabsBrowserTabValue

export type SessionTabsTabGroupValue = {
  id: string
  activeTabId: string | null
  tabOrder: string[]
  recentTabIds?: string[]
}

export type SessionTabsGroupLayoutNodeValue =
  | { type: 'leaf'; groupId: string }
  | {
      type: 'split'
      direction: 'horizontal' | 'vertical'
      first: SessionTabsGroupLayoutNodeValue
      second: SessionTabsGroupLayoutNodeValue
      ratio?: number
    }

export type SessionTabsSnapshotValue = {
  worktree: string
  publicationEpoch: string
  snapshotVersion: number
  activeGroupId: string | null
  activeTabId: string | null
  activeTabType: SessionTabsTabTypeValue | null
  tabGroups?: SessionTabsTabGroupValue[]
  tabGroupLayout?: SessionTabsGroupLayoutNodeValue | null
  tabs: SessionTabsTabValue[]
  removed?: true
}

export type SessionTabsCreateTerminalResultValue = {
  tab: SessionTabsTerminalTabValue
  publicationEpoch: string
  snapshotVersion: number
}

export type SessionTabsStreamEventValue =
  | ({ type: 'snapshot' } & SessionTabsSnapshotValue)
  | ({ type: 'updated' } & SessionTabsSnapshotValue)
  | { type: 'end' }

export type SessionTabsAllStreamEventValue =
  | { type: 'snapshots'; snapshots: SessionTabsSnapshotValue[] }
  | ({ type: 'updated' } & SessionTabsSnapshotValue)
  | { type: 'end' }

export type SessionTabsStartupCommandDeliveryInput = 'fast' | 'shell-ready'

export type SessionTabsCreateTerminalInput = {
  worktree: string
  activate?: boolean
  afterTabId?: string
  agent?: string
  agentPrompt?: string
  clientMutationId?: string
  command?: string
  cwd?: string
  env?: Record<string, string>
  envToDelete?: string[]
  launchAgent?: string
  launchConfig?: {
    agentCommand?: string
    agentArgs: string
    agentEnv: Record<string, string>
    ompResumeFilePath?: string
  }
  launchToken?: string
  startupCommandDelivery?: SessionTabsStartupCommandDeliveryInput
  targetGroupId?: string
}

export type SessionTabsMoveInput = {
  worktree: string
  tabId: string
  targetGroupId: string
  kind: 'reorder' | 'move-to-group' | 'split'
  tabOrder?: string[]
  index?: number
  splitDirection?: 'left' | 'right' | 'up' | 'down'
}

export type SessionTabsSetTabPropsInput = {
  worktree: string
  tabId: string
  color?: string | null
  isPinned?: boolean
}

export type SessionTabsPaneLayoutNodeValue =
  | { type: 'leaf'; leafId: string }
  | {
      type: 'split'
      direction: 'horizontal' | 'vertical'
      first: SessionTabsPaneLayoutNodeValue
      second: SessionTabsPaneLayoutNodeValue
      ratio?: number
    }

export type SessionTabsUpdatePaneLayoutInput = {
  worktree: string
  tabId: string
  root: SessionTabsPaneLayoutNodeValue | null
  expandedLeafId: string | null
  titlesByLeafId?: Record<string, string>
}

File: apps/mobile/YiruMobile/Platform/Runtime/SessionTabs/Values.swift
import Foundation

nonisolated enum MobileSessionTabTypeWire: String, Decodable, Sendable {
    case terminal = "terminal"
    case markdown = "markdown"
    case file = "file"
    case browser = "browser"
}

nonisolated enum MobileSessionTerminalStatusWire: String, Decodable, Sendable {
    case pendingHandle = "pending-handle"
    case ready = "ready"
    case sleeping = "sleeping"
}

nonisolated struct MobileSessionCreateTerminalRequestWire: Encodable, Sendable {
    let worktree: String
    let afterTabId: String?
    let activate: Bool?
    let clientMutationId: String?
    let agent: String?
    let command: String?
    let env: [String: String]?
    let envToDelete: [String]?
    let launchConfig: MobileSleepingAgentLaunchConfigWire?
    let launchAgent: String?
    let startupCommandDelivery: String?
    let agentPrompt: String?
}

nonisolated struct MobileSleepingAgentLaunchConfigWire: Encodable, Sendable {
    let agentCommand: String?
    let agentArgs: String
    let agentEnv: [String: String]
    let ompResumeFilePath: String?
}

nonisolated struct MobileSessionProviderSessionWire: Decodable, Sendable {
    let key: String
    let id: String
    let transcriptPath: String?
}

nonisolated struct MobileSessionAgentStatusWire: Decodable, Sendable {
    let state: String
    let paneKey: String?
    let prompt: String?
    let updatedAt: Double?
    let stateStartedAt: Double?
    let agentType: String?
    let interactivePrompt: String?
    let lastAssistantMessage: String?
    let toolName: String?
    let toolInput: String?
    let interrupted: Bool?
    let providerSession: MobileSessionProviderSessionWire?
}

nonisolated struct MobileSessionTabWire: Decodable, Sendable {
    let id: String
    let title: String
    let isActive: Bool
    let color: String?
    let isPinned: Bool?
    let type: MobileSessionTabTypeWire
    let parentTabId: String?
    let leafId: String?
    let ptyId: String?
    let launchAgent: String?
    let resolvedAgentType: String?
    let agentStatus: MobileSessionAgentStatusWire?
    let status: MobileSessionTerminalStatusWire?
    let terminal: String?
    let worktreeInstanceId: String?
    let filePath: String?
    let relativePath: String?
    let language: String?
    let mode: String?
    let diffSource: String?
    let isDirty: Bool?
    let sourceFileId: String?
    let sourceFilePath: String?
    let sourceRelativePath: String?
    let documentVersion: String?
    let browserWorkspaceId: String?
    let browserPageId: String?
    let url: String?
    let loading: Bool?
    let canGoBack: Bool?
    let canGoForward: Bool?
}

nonisolated struct MobileSessionTabsWire: Decodable, Sendable {
    let worktree: String
    let publicationEpoch: String
    let snapshotVersion: Int64
    let activeTabId: String?
    let activeTabType: MobileSessionTabTypeWire?
    let tabs: [MobileSessionTabWire]
}

nonisolated struct MobileSessionCreateTerminalResultWire: Decodable, Sendable {
    let tab: MobileSessionTabWire
    let publicationEpoch: String
    let snapshotVersion: Int64
}

nonisolated enum MobileSessionTabsEventWire: Decodable, Sendable {
    case snapshot(MobileSessionTabsWire)
    case updated(MobileSessionTabsWire)
    case end

    private enum EventType: String, Decodable {
        case snapshot
        case updated
        case end
    }

    private enum CodingKeys: String, CodingKey {
        case type
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        switch try container.decode(EventType.self, forKey: .type) {
        case .snapshot:
            self = .snapshot(try MobileSessionTabsWire(from: decoder))
        case .updated:
            self = .updated(try MobileSessionTabsWire(from: decoder))
        case .end:
            self = .end
        }
    }
}

nonisolated enum MobileSessionTabsAllEventWire: Decodable, Sendable {
    case snapshots([MobileSessionTabsWire])
    case updated(MobileSessionTabsWire)
    case end

    private enum EventType: String, Decodable {
        case snapshots
        case updated
        case end
    }

    private enum CodingKeys: String, CodingKey {
        case type
        case snapshots
    }

    init(from decoder: any Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        switch try container.decode(EventType.self, forKey: .type) {
        case .snapshots:
            self = .snapshots(
                try container.decode([MobileSessionTabsWire].self, forKey: .snapshots)
            )
        case .updated:
            self = .updated(try MobileSessionTabsWire(from: decoder))
        case .end:
            self = .end
        }
    }
}
